### ISSUE TITLE
Builder

### ISSUE THREAD

**Issue by MohamedAnwarAlhamed** [MohamedAnwarAlhamed/CSharp_Design_Patterns#2]
The Builder Design Pattern is a creational design pattern used to construct complex objects step by step. Unlike other creational patterns, the Builder pattern allows the creation of different representations of an object while hiding the construction logic from the client. It is particularly useful when an object requires multiple initialization steps, especially when some of these steps are optional or when the object being built can vary in its final form.

Intent
-
- Separate the construction of a complex object from its representation, so that the same construction process can create different representations of the object.
- Provide a flexible solution to assembling objects with various configurations.
- Allow step-by-step construction of an object.

Problem
-
Creating an object with many parameters or requiring several steps can result in complex, hard-to-maintain code, especially if the object has optional parameters or different configurations. Using telescoping constructors (constructors with varying numbers of parameters) can make the code less readable and difficult to manage.

For example, if you're building a house, you may want to customize it by specifying the number of rooms, windows, doors, etc. Without a structured approach like the Builder pattern, you'd need to overload constructors or pass many parameters, which becomes unwieldy.

Solution
-
The Builder pattern provides a way to construct an object step by step. The construction process can be customized by altering the steps, which allows for flexible creation of objects in various configurations. The pattern is especially useful for constructing immutable objects or objects that require complex assembly.

The Builder Design Pattern is a creational design pattern used to construct complex objects step by step. Unlike other creational patterns, the Builder pattern allows the creation of different representations of an object while hiding the construction logic from the client. It is particularly useful when an object requires multiple initialization steps, especially when some of these steps are optional or when the object being built can vary in its final form.

Structure
-
The Builder pattern consists of the following components:
- Product: The complex object being constructed. This class defines the object that will be created by the builder.
- Builder: This interface defines the steps (methods) required to construct the product. Each step typically corresponds to setting a part of the product.
- Concrete Builder: A class that implements the Builder interface. It defines how to perform each step in the construction process and assembles the product.
- Director: This class controls the construction process. It is responsible for executing the steps defined by the Builder in a specific sequence to create the product. The director does not know the specifics of the product being built.
- Client: The client creates a Director and a Concrete Builder. The client instructs the director to build the product, step by step.

Define the Product (House)
```c#
// Product
public class House
{
    public string WallType { get; set; }
    public string DoorType { get; set; }
    public string WindowType { get; set; }
    public string RoofType { get; set; }

    public void ShowHouse()
    {
        Console.WriteLine($"House with {WallType} walls, {DoorType} doors, {WindowType} windows, and a {RoofType} roof.");
    }
}

```

Define the Builder Interface
```c#
// Builder Interface
public interface IHouseBuilder
{
    void BuildWalls();
    void BuildDoors();
    void BuildWindows();
    void BuildRoof();
    House GetHouse();
}

```
Define Concrete Builders
```c#
// Concrete Builder for Wooden House
public class WoodenHouseBuilder : IHouseBuilder
{
    private House house = new House();

    public void BuildWalls() => house.WallType = "Wooden";
    public void BuildDoors() => house.DoorType = "Wooden";
    public void BuildWindows() => house.WindowType = "Wooden";
    public void BuildRoof() => house.RoofType = "Wooden";

    public House GetHouse() => house;
}

// Concrete Builder for Glass House
public class GlassHouseBuilder : IHouseBuilder
{
    private House house = new House();

    public void BuildWalls() => house.WallType = "Glass";
    public void BuildDoors() => house.DoorType = "Glass";
    public void BuildWindows() => house.WindowType = "Glass";
    public void BuildRoof() => house.RoofType = "Glass";

    public House GetHouse() => house;
}

```

Define the Director
```c#
// Director
public class ConstructionDirector
{
    public void ConstructHouse(IHouseBuilder houseBuilder)
    {
        houseBuilder.BuildWalls();
        houseBuilder.BuildDoors();
        houseBuilder.BuildWindows();
        houseBuilder.BuildRoof();
    }
}

```

Client Code
```c#
class Program
{
    static void Main(string[] args)
    {
        // Create the director
        ConstructionDirector director = new ConstructionDirector();

        // Build a wooden house
        IHouseBuilder woodenHouseBuilder = new WoodenHouseBuilder();
        director.ConstructHouse(woodenHouseBuilder);
        House woodenHouse = woodenHouseBuilder.GetHouse();
        woodenHouse.ShowHouse(); // Output: House with Wooden walls, Wooden doors, Wooden windows, and a Wooden roof.

        // Build a glass house
        IHouseBuilder glassHouseBuilder = new GlassHouseBuilder();
        director.ConstructHouse(glassHouseBuilder);
        House glassHouse = glassHouseBuilder.GetHouse();
        glassHouse.ShowHouse(); // Output: House with Glass walls, Glass doors, Glass windows, and a Glass roof.
    }
}

```

Advantages of the Builder Pattern
-
- Simplifies Object Creation: The pattern makes it easy to create complex objects step by step and control their configuration.
- Readable Code: You avoid a bloated constructor with many parameters, improving readability and maintainability.
- Reusability: The same construction process can be reused to build different types of products with slight variations.
- Encapsulation of Object Creation: The object creation logic is encapsulated in the builder, making the client code cleaner.
- Flexible and Extensible: New types of products can be introduced by adding new concrete builders without changing existing client code.

Disadvantages of the Builder Pattern
-
- More Complex: It can introduce extra classes (builders and directors), leading to more complexity for simple object creation.
- Overhead: In cases where object construction is simple, the Builder pattern may introduce unnecessary overhead.

When to Use the Builder Design Pattern
- When constructing a complex object: If an object requires multiple steps to initialize, the Builder pattern can simplify the construction process.
- When object creation is highly configurable: If the object being constructed can have a variety of configurations or optional parameters, the Builder pattern is ideal.
- When immutability is required: Builders can help create immutable objects step by step without requiring multiple constructors or setter methods.
- When you want to separate concerns: The builder pattern is useful when you want to separate the object construction process from the object's business logic or representation.

Real-World Examples of the Builder Pattern
-
- StringBuilder in C#: The StringBuilder class in C# allows you to build strings in a step-by-step fashion, adding elements to the string without the need for creating multiple intermediate objects.
- HTML Builders: Building complex HTML documents or user interfaces can benefit from the Builder pattern, where each element (like headers, paragraphs, divs) is added in a controlled manner.
- SQL Query Builders: Many database libraries use builders to construct complex SQL queries step by step, allowing different parts of the query (select, where, join clauses) to be added separately.

In conclusion, the Builder Design Pattern is a robust and flexible solution for constructing complex objects step by step. It promotes clean code, encapsulates the construction logic, and provides a clear way to build objects with different configurations while maintaining good separation of concerns.
